dsp/impl/magnitude_u16o12.c

```diff
@@ -0,0 +1,189 @@
+#include <stdlib.h>
+#include <math.h>
+#include <endian.h>
+#include "dsp/helpers/tables.h"
+
+/*
+ * Convert (little-endian) unsigned 16 offset 12 ( Excess 2048 format)
+ * values to unsigned 16-bit magnitudes.
+ *
+ * Offset 12:
+ *
+ *          Signed      Scaled
+ * Input    Value       Magnitude
+ * ------------------------------
+ * 0        <invalid>   <invalid>
```

Comment:
If this is raw ADC output, then 0 is probably a valid value, let's handle it.

You could either shift the whole scale by 0.5 (i.e. ADC value of 2047.5 is magnitude 0), or saturate input 0 -> magnitude 65535, or tweak the scale factor slightly such that it doesn't actually reach 65535 at both ends.

Reply:
I've been testing for that 0 for months now and haven't run across it and the definition of the offset representation specifically says it's invalid.  I chose the scaling I did so that a max negative (-2047) and a max positive (2047) became 65535 on output.  For the lookup table, I can add a value in the 0 slot.  The calculation will simply return a rolled over value for that sample which should be no big deal.